ml_metadata/tools/mlmd_bench/thread_runner.cc

```diff
@@ -0,0 +1,114 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "ml_metadata/tools/mlmd_bench/thread_runner.h"
+
+#include <vector>
+
+#include "absl/synchronization/mutex.h"
+#include "ml_metadata/metadata_store/metadata_store.h"
+#include "ml_metadata/metadata_store/metadata_store_factory.h"
+#include "ml_metadata/metadata_store/types.h"
+#include "ml_metadata/tools/mlmd_bench/benchmark.h"
+#include "ml_metadata/tools/mlmd_bench/proto/mlmd_bench.pb.h"
+#include "ml_metadata/tools/mlmd_bench/stats.h"
+#include "ml_metadata/tools/mlmd_bench/workload.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/lib/core/threadpool.h"
+
+namespace ml_metadata {
+
+ThreadRunner::ThreadRunner(const MLMDBenchConfig& mlmd_bench_config)
+    : num_threads_(mlmd_bench_config.thread_env_config().num_threads()),
+      mlmd_config_(mlmd_bench_config.mlmd_config()) {}
+
+// Execution unit of mlmd_bench. Returns detailed error if query executions
+// failed.
+// The thread runner will first loops over all the executable workloads in
+// benchmark and executes them one by one. Each workload will have a
+// `thread_stats_list` to record the stats of each thread when executing the
+// current workload.
+// During the execution, each operation will has a `op_stats` to record current
+// operation statistic. Each `op_stats` will be used to update the
+// `thread_stats`.
+// After the each thread has finished the execution, the workload stats will be
+// generated by merging all the thread stats inside the `thread_stats_list`. The
+// performance of the workload will be reported according to the workload stats.
+tensorflow::Status ThreadRunner::Run(Benchmark& benchmark) {
+  for (auto& workload : benchmark.workloads()) {
+    Stats thread_stats_list[num_threads_];
+    // `workload.second` is the number of operations for current executable 
+    // workload.
+    int64 op_per_thread = workload.second / num_threads_;
+    std::unique_ptr<MetadataStore> set_up_store;
+    TF_RETURN_IF_ERROR(CreateMetadataStore(mlmd_config_, &set_up_store));
+    // `workload.first` is the executable workload.
+    TF_RETURN_IF_ERROR(workload.first->SetUp(set_up_store.get()));
+    {
+      // Create a thread pool for multi-thread execution.
+      tensorflow::thread::ThreadPool pool(tensorflow::Env::Default(),
+                                          "mlmd_bench", num_threads_);
+      // `total_done` is used for reporting progress along the way.
+      int64 total_done = 0;
+      int64 thread_index = 0;
+      for (int64 i = 0; i < num_threads_; ++i) {
+        pool.Schedule([&]() {
+          // Uses a lock to ensure the consistence of concurrent reading and
+          // writing of `thread`.
+          lock_.Lock();
+          int64 curr_thread_index = thread_index;
```

Comment:
instead of using a thread_index, and lock/unlock to increase it, you can simply pass the thread_stat_list[i] to the Schedule lambda, each thread will then use its own stats. also instead of a default capture, use explicit capture of the variable needed in the lambda, see style guide [1]

[1] https://google.github.io/styleguide/cppguide.html#Lambda_expressions

Reply:
Yeah, this is better! Fixed! Thanks :)